src/v/cluster/node_status_provider_rpc_handler.cc

```diff
@@ -0,0 +1,20 @@
+#include "cluster/node_status_provider_rpc_handler.h"
+
+namespace cluster {
+
+node_status_provider_rpc_handler::node_status_provider_rpc_handler(
+  ss::scheduling_group sg,
+  ss::smp_service_group ssg,
+  node_status_provider& node_status_provider)
+  : node_status_provider_rpc_service(sg, ssg)
+  , _node_status_provider(node_status_provider) {}
+
+ss::future<node_status_reply> node_status_provider_rpc_handler::node_status(
+  node_status_request&& r, rpc::streaming_context&) {
+    return ss::smp::submit_to(
+      node_status_provider::shard, [this, r = std::move(r)]() mutable {
+          return _node_status_provider.process_request(std::move(r));
+      });
```

Comment:
i think we should prefer to use the `ss::sharded<T>::invoke_on` abstraction by default and reserve the low-level cross-core messaging (like smp::submit_to) for cases that don't fit into the sharded service model. is this an exception?

Reply:
I considered it, but I expected an object of type `sharded<T>` to operate on multiple cores. `node_status_provider` runs only on one core and we'd be wrapping it in `sharded` only to be able to dispatch to core 0. Feels a bit awkward to me.

Reply:
> node_status_provider::shard,

you are right. for some reason i was thinking that this was jumping to the shard with the target connection from the connection cache.

still i think it can be avoided. consider the case in which you want to replicate the metadata about node status on each core (likely). in this case, you'd still have a sharded service maintaining that state. when handling an rpc you can jump to a home work via sharded::invoke_on and then broadcast the state to the other cores.

the reason we try to avoid submit_to is that it brings along some complications that are useful to not have to consider https://github.com/scylladb/seastar/blob/master/include/seastar/core/smp.hh#L83. in this particular case it doesn't look like there is anything forcing the use of submit_to.

@mmaslankaprv any tips on how to organize the services to avoid raw submit_to? does the controller service manage to avoid raw submit_to calls?

Reply:
We have multiple examples of situations in which we only create a single instance of `ss::sharded<T>` service. (f.e. in `controller.h`). I think it is much better to use sharded service `start_single_instance` as it will assert out when one would try to access instances that does not exists. Additionally it will force caller to deal with the `sharded `interface hence it will force it to check if x-core call is required. Plain `std::unique_ptr` has none of those semantics embedded. it is very easy to call it without the `ss::submit_to` and this may lead to some subtle intermittent concurrency bugs.

Reply:
That makes sense. I'll make the service sharded.

Noah also makes a good point about replicating the metadata on all cores. There's a trade-off here. Replicating every update to all cores means having a lot of cross-shard writes, but no cross-shard reads. If we don't replicate, every read will have to jump shards. We are going to have significantly less reads than writes so maybe keeping the metadata on a single shard is preferable?

We could also batch updates somehow, but that introduces delays and complicates the semantics so I'm not too in favour of it.

Reply:
> Replicating every update to all cores means having a lot of cross-shard writes, but no cross-shard reads

in the best case scenario shards should be able to do all their work without any cross-core communication. in reality, the goal is to minimize it. the metadata cache and leadership dissementation service are good examples: reads are cheap, and occasional updates keep all the cores in sync.

you can debounce global updates across cores if needed.

Reply:
Concerning replication of node status metadata across cores there's three options:
1. Replicate on every status update (whenever a node_status request or reply are received)
2. Batch updates for other cores
3. Keep the data on one shard and do cross-shard reads

For the first iteration I think option 1 is sensible. For option 2, keep in mind that in order for Raft to be able to use the timestamp from `node_status_provider` it needs to be smaller than the configured raft timeout (otherwise it doesn't add any confidence that the peer may be in a good state). This imposes a hard time limit on batch flushing and makes it flaky. Option 3 would mean a constant stream of x-shard reads, and I don't think we want to pay this baseline cost to optimise reads right now.

Reply:
While working on the Raft integration I've changed my mind and replicated the data across all shards. See the latest force push.